File: Pandacap.LowLevel/Data/PandacapDbContext.fs
namespace Pandacap.Data

open Microsoft.EntityFrameworkCore

type PandacapDbContext(options: DbContextOptions<PandacapDbContext>) =
    inherit DbContext(options)

    member this.DeviantArtCredentials: DbSet<DeviantArtCredentials> = this.Set()
    member this.FurAffinityCredentials: DbSet<FurAffinityCredentials> = this.Set()
    member this.RedditCredentials: DbSet<RedditCredentials> = this.Set()
    member this.WeasylCredentials: DbSet<WeasylCredentials> = this.Set()
    member this.Avatars: DbSet<Avatar> = this.Set()
    member this.Posts: DbSet<Post> = this.Set()
    member this.Uploads: DbSet<Upload> = this.Set()
    member this.AddressedPosts: DbSet<AddressedPost> = this.Set()
    member this.PostActivities: DbSet<PostActivity> = this.Set()
    member this.DeviantArtTextPostCheckStatuses: DbSet<DeviantArtTextPostCheckStatus> = this.Set()
    member this.InboxArtworkDeviations: DbSet<InboxArtworkDeviation> = this.Set()
    member this.InboxTextDeviations: DbSet<InboxTextDeviation> = this.Set()
    member this.InboxActivityStreamsPosts: DbSet<InboxActivityStreamsPost> = this.Set()
    member this.InboxFurAffinityJournals: DbSet<InboxFurAffinityJournal> = this.Set()
    member this.InboxFurAffinitySubmissions: DbSet<InboxFurAffinitySubmission> = this.Set()
    member this.InboxWeasylJournals: DbSet<InboxWeasylJournal> = this.Set()
    member this.InboxWeasylSubmissions: DbSet<InboxWeasylSubmission> = this.Set()
    member this.GeneralFeeds: DbSet<GeneralFeed> = this.Set()
    member this.GeneralInboxItems: DbSet<GeneralInboxItem> = this.Set()
    member this.GeneralFavorites: DbSet<GeneralFavorite> = this.Set()
    member this.Follows: DbSet<Follow> = this.Set()
    member this.Followers: DbSet<Follower> = this.Set()
    member this.CommunityBookmarks: DbSet<CommunityBookmark> = this.Set()
    member this.ATProtoFeeds: DbSet<ATProtoFeed> = this.Set()
    member this.BlueskyLikeFeedItems: DbSet<BlueskyLikeFeedItem> = this.Set()
    member this.BlueskyPostFeedItems: DbSet<BlueskyPostFeedItem> = this.Set()
    member this.BlueskyRepostFeedItems: DbSet<BlueskyRepostFeedItem> = this.Set()
    member this.WhiteWindBlogEntryFeedItems: DbSet<WhiteWindBlogEntryFeedItem> = this.Set()
    member this.LeafletDocumentFeedItems: DbSet<LeafletDocumentFeedItem> = this.Set()
    member this.RemoteActivityPubAddressedPosts: DbSet<RemoteActivityPubAddressedPost> = this.Set()
    member this.RemoteActivityPubReplies: DbSet<RemoteActivityPubReply> = this.Set()
    member this.ActivityPubOutboundActivities: DbSet<ActivityPubOutboundActivity> = this.Set()
    member this.ActivityPubFavorites: DbSet<ActivityPubFavorite> = this.Set()
    member this.BlueskyPostFavorites: DbSet<BlueskyPostFavorite> = this.Set()
    member this.DeviantArtFavorites: DbSet<DeviantArtFavorite> = this.Set()
    member this.FurAffinityFavorites: DbSet<FurAffinityFavorite> = this.Set()
    member this.RedditUpvotedPosts: DbSet<RedditUpvotedPost> = this.Set()
    member this.WeasylFavoriteSubmissions: DbSet<WeasylFavoriteSubmission> = this.Set()

    override _.OnModelCreating(builder) =
        base.OnModelCreating(builder)
        ignore [
            builder.HasEmbeddedDiscriminatorName("Discriminator")
            builder.HasDiscriminatorInJsonIds()
            builder.HasShadowIds()
        ]

File: Pandacap.Database/PandacapDbContext.cs
using Microsoft.EntityFrameworkCore;
using System.Text.Json;

namespace Pandacap.Database
{
    public interface IOfflinePlatformDataCache
    {
        Task<T?> TryGetAsync<T>(
            OfflinePlatformDataCacheItem.CachedPlatformDataType type,
            CancellationToken cancellationToken = default) where T : class;

        Task UpdateAsync<T>(
            OfflinePlatformDataCacheItem.CachedPlatformDataType type,
            T value,
            CancellationToken cancellationToken = default) where T : class;
    }

    public class PandacapDbContext(DbContextOptions<PandacapDbContext> options) : DbContext(options), IOfflinePlatformDataCache
    {
        public DbSet<DeviantArtCredentials> DeviantArtCredentials { get; set; }
        public DbSet<Avatar> Avatars { get; set; }
        public DbSet<Post> Posts { get; set; }
        public DbSet<Upload> Uploads { get; set; }
        public DbSet<AddressedPost> AddressedPosts { get; set; }
        public DbSet<PostActivity> PostActivities { get; set; }
        public DbSet<DeviantArtTextPostCheckStatus> DeviantArtTextPostCheckStatuses { get; set; }
        public DbSet<InboxArtworkDeviation> InboxArtworkDeviations { get; set; }
        public DbSet<InboxTextDeviation> InboxTextDeviations { get; set; }
        public DbSet<InboxActivityStreamsPost> InboxActivityStreamsPosts { get; set; }
        public DbSet<InboxFurAffinityJournal> InboxFurAffinityJournals { get; set; }
        public DbSet<InboxFurAffinitySubmission> InboxFurAffinitySubmissions { get; set; }
        public DbSet<InboxWeasylJournal> InboxWeasylJournals { get; set; }
        public DbSet<InboxWeasylSubmission> InboxWeasylSubmissions { get; set; }
        public DbSet<GeneralFeed> GeneralFeeds { get; set; }
        public DbSet<GeneralInboxItem> GeneralInboxItems { get; set; }
        public DbSet<Follow> Follows { get; set; }
        public DbSet<Follower> Followers { get; set; }
        public DbSet<CommunityBookmark> CommunityBookmarks { get; set; }
        public DbSet<ATProtoFeed> ATProtoFeeds { get; set; }
        public DbSet<BlueskyLikeFeedItem> BlueskyLikeFeedItems { get; set; }
        public DbSet<BlueskyPostFeedItem> BlueskyPostFeedItems { get; set; }
        public DbSet<BlueskyRepostFeedItem> BlueskyRepostFeedItems { get; set; }
        public DbSet<StandardSiteDocumentFeedItem> StandardSiteDocumentFeedItems { get; set; }
        public DbSet<RemoteActivityPubAddressedPost> RemoteActivityPubAddressedPosts { get; set; }
        public DbSet<RemoteActivityPubReply> RemoteActivityPubReplies { get; set; }
        public DbSet<ActivityPubOutboundActivity> ActivityPubOutboundActivities { get; set; }
        public DbSet<ActivityPubFavorite> ActivityPubFavorites { get; set; }
        public DbSet<BlueskyPostFavorite> BlueskyPostFavorites { get; set; }
        public DbSet<DeviantArtFavorite> DeviantArtFavorites { get; set; }
        public DbSet<FurAffinityFavorite> FurAffinityFavorites { get; set; }
        public DbSet<WeasylFavoriteSubmission> WeasylFavoriteSubmissions { get; set; }
        public DbSet<CanonicalMedium> CanonicalMediums { get; set; }
        public DbSet<CanonicalSetting> CanonicalSettings { get; set; }
        public DbSet<CanonicalCharacter> CanonicalCharacters { get; set; }
        public DbSet<CanonicalSpecies> CanonicalSpecies { get; set; }
        public DbSet<OfflinePlatformDataCacheItem> CachedPlatformData { get; set; }
        public DbSet<FurAffinityNotification> FurAffinityNotifications { get; set; }
        public DbSet<FurAffinityNote> FurAffinityNotes { get; set; }
        public DbSet<WeasylNotification> WeasylNotifications { get; set; }
        public DbSet<WeasylNote> WeasylNotes { get; set; }

        public IOfflinePlatformDataCache OfflinePlatformDataCache => this;

        async Task<T?> IOfflinePlatformDataCache.TryGetAsync<T>(
            OfflinePlatformDataCacheItem.CachedPlatformDataType type,
            CancellationToken cancellationToken) where T : class
        {
            var platformData = await CachedPlatformData
                .Where(d => d.Type == type)
                .SingleOrDefaultAsync(cancellationToken);

            if (platformData == null)
                return null;

            try
            {
                return JsonSerializer.Deserialize<T?>(platformData.Json);
            }
            catch (Exception)
            {
                return null;
            }
        }

        async Task IOfflinePlatformDataCache.UpdateAsync<T>(
            OfflinePlatformDataCacheItem.CachedPlatformDataType type,
            T value,
            CancellationToken cancellationToken) where T : class
        {
            var platformData = await CachedPlatformData
                .Where(d => d.Type == type)
                .SingleOrDefaultAsync(cancellationToken);

            if (platformData == null)
            {
                platformData = new() { Type = type };
                CachedPlatformData.Add(platformData);
            }

            platformData.Json = JsonSerializer.Serialize(value);

            await SaveChangesAsync(cancellationToken);
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            base.OnModelCreating(modelBuilder);
            modelBuilder.HasEmbeddedDiscriminatorName("Discriminator");
            modelBuilder.HasDiscriminatorInJsonIds();
            modelBuilder.HasShadowIds();
        }
    }
}
